[Testerina] Refactor exception handling in Test Framework

**Description:**
Exception handling in the Test Framework needs to be handled taking various scenarios into account

- [ ] Throw an exception when module init of a imported module
- [x] Throw an exception from interop code
- [ ] Class not found in module init of a imported module
- [x] Class not found in test running
- [x] Class def not found for imported dependencies
- [ ] Class cast when running test function
- [x] Class cast when running module init of a imported module
- [ ] BadSad from user code

**Describe your problem(s)**

**Describe your solution(s)**

**Related Issues (optional):**
https://github.com/ballerina-platform/ballerina-lang/issues/28151
<!-- Any related issues such as sub tasks, issues reported in other repositories (e.g component repositories), similar problems, etc. -->

**Suggested Labels (optional):**
<!-- Optional comma separated list of suggested labels. Non committers can’t assign labels to issues, so this will help issue creators who are not a committer to suggest possible labels-->

**Suggested Assignees (optional):**
<!--Optional comma separated list of suggested team members who should attend the issue. Non committers can’t assign issues to assignees, so this will help issue creators who are not a committer to suggest possible assignees-->

Class cast in log module

Stack 
```
[2021-03-10 09:28:44,406] SEVERE {b7a.log.crash} - Dependant module initialization for test suite failed due to java.lang.ExceptionInInitializerError 
org.ballerinalang.test.runtime.exceptions.BallerinaTestException: Dependant module initialization for test suite failed due to java.lang.ExceptionInInitializerError
	at org.ballerinalang.test.runtime.BTestRunner.startSuite(BTestRunner.java:547)
	at org.ballerinalang.test.runtime.BTestRunner.execute(BTestRunner.java:265)
	at org.ballerinalang.test.runtime.BTestRunner.runTest(BTestRunner.java:115)
	at org.ballerinalang.test.runtime.util.TesterinaUtils.executeTests(TesterinaUtils.java:76)
	at org.ballerinalang.test.runtime.Main.startTestSuit(Main.java:115)
	at org.ballerinalang.test.runtime.Main.main(Main.java:96)
Caused by: java.lang.ExceptionInInitializerError
	at ballerina.log.1_1_0-alpha5.natives.setGlobalLogLevelExtern(natives.bal:174)
	at ballerina.log.1_1_0-alpha5.natives.initializeLogLevels(natives.bal:168)
	at ballerina.log.1_1_0-alpha5.init.init(init.bal:21)
	at ballerina.log.1_1_0-alpha5.$_init.$gen$$0046$0060init$0062(log:1)
	at ballerina.log.1_1_0-alpha5.$_init.$moduleInit(log)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.ballerinalang.test.runtime.entity.TesterinaFunction.lambda$runOnSchedule$0(TesterinaFunction.java:109)
	at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(Scheduler.java:571)
	at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:301)
	at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:269)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassCastException: class java.util.logging.LogManager cannot be cast to class org.ballerinalang.logging.BLogManager (java.util.logging.LogManager is in module java.logging of loader 'bootstrap'; org.ballerinalang.logging.BLogManager is in unnamed module of loader 'app')
	at org.ballerinalang.stdlib.log.AbstractLogFunction.<clinit>(AbstractLogFunction.java:38)
	... 14 more
 ```

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java
@@ -38,6 +38,7 @@ import io.ballerina.runtime.api.values.BError;
 import io.ballerina.runtime.api.values.BString;
 import io.ballerina.runtime.internal.scheduling.Scheduler;
 import io.ballerina.runtime.internal.scheduling.Strand;
+import io.ballerina.runtime.internal.util.RuntimeUtils;
 import io.ballerina.runtime.internal.values.ArrayValue;
 import io.ballerina.runtime.internal.values.DecimalValue;
 import io.ballerina.runtime.internal.values.MapValue;
@@ -426,7 +427,7 @@ public class BTestRunner {
                 functionResult = new TesterinaResult(test.getTestName(), false, true, null);
                 tReport.addFunctionResult(packageName, functionResult);
             }
-        } catch (Throwable e) {
+        } catch (BallerinaTestException | ClassNotFoundException e) {
             // If the test function is skipped lets add it to the failed test list
             failedOrSkippedTests.add(test.getTestName());
             // report the test result
@@ -626,7 +627,12 @@ public class BTestRunner {
                 throw new BallerinaTestException(e);
             }
         } else if (e instanceof BallerinaTestException) {
-            throw (BallerinaTestException) e;
+            try {
+                message = TesterinaUtils.getPrintableStackTrace(e.getCause());
+                RuntimeUtils.silentlyLogBadSad(e.getCause());
+            } catch (ClassCastException castException) {
+                throw new BallerinaTestException(e);
+            }
         } else {
             throw new BallerinaTestException(e);
         }

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java
@@ -405,19 +405,15 @@ public class BTestRunner {
                     valueSets = invokeTestFunction(suite, test.getDataProvider(), classLoader, scheduler);
                 }
                 if (valueSets == null) {
-                    invokeTestFunction(suite, test.getTestName(), classLoader, scheduler);
-                    // report the test result
-                    functionResult = new TesterinaResult(test.getTestName(), true, shouldSkip
-                            .get(), null);
-                    tReport.addFunctionResult(packageName, functionResult);
+                    valueSets = invokeTestFunction(suite, test.getTestName(), classLoader, scheduler);
+                    computeFunctionResult(test, packageName, shouldSkip, failedOrSkippedTests, valueSets);
                 } else {
                     Class<?>[] argTypes = extractArgumentTypes(valueSets);
                     List<Object[]> argList = extractArguments(valueSets);
                     for (Object[] arg : argList) {
-                        invokeTestFunction(suite, test.getTestName(), classLoader, scheduler, argTypes, arg);
-                        TesterinaResult result = new TesterinaResult(test.getTestName(), true,
-                                                                     shouldSkip.get(), null);
-                        tReport.addFunctionResult(packageName, result);
+                        valueSets = invokeTestFunction(suite, test.getTestName(), classLoader, scheduler, argTypes,
+                                arg);
+                        computeFunctionResult(test, packageName, shouldSkip, failedOrSkippedTests, valueSets);
                     }
                 }
             } else {
@@ -446,6 +442,30 @@ public class BTestRunner {
             writeFailedTestsToJson(failedOrSkippedTests, jsonFile);
         }
 
+    }
+
+    private void computeFunctionResult(Test test, String packageName, AtomicBoolean shouldSkip,
+                                       List<String> failedOrSkippedTests, Object valueSets) {
+        TesterinaResult functionResult;
+        if (valueSets instanceof BError) {
+            failedOrSkippedTests.add(test.getTestName());
+            functionResult = new TesterinaResult(test.getTestName(), false, shouldSkip.get(),
+                    formatErrorMessage((BError) valueSets));
+            tReport.addFunctionResult(packageName, functionResult);
+        } else if (valueSets instanceof Exception) {
+            failedOrSkippedTests.add(test.getTestName());
+            functionResult = new TesterinaResult(test.getTestName(), false, shouldSkip.get(),
+                    formatErrorMessage((Exception) valueSets));
+            tReport.addFunctionResult(packageName, functionResult);
+        } else if (valueSets instanceof Error) {
+            failedOrSkippedTests.add(test.getTestName());
+            functionResult = new TesterinaResult(test.getTestName(), false, shouldSkip.get(),
+                    formatErrorMessage((Error) valueSets));
+            tReport.addFunctionResult(packageName, functionResult);
+        } else {
+            functionResult = new TesterinaResult(test.getTestName(), true, shouldSkip.get(), null);
+            tReport.addFunctionResult(packageName, functionResult);
+        }
     }
 
     private void executeAfterFunction(Test test, TestSuite suite, ClassLoader classLoader, Scheduler scheduler,
@@ -610,32 +630,30 @@ public class BTestRunner {
         return testerinaFunction.invoke();
     }
 
-    public void invokeTestFunction(TestSuite suite, String functionName, ClassLoader classLoader,
+    public Object invokeTestFunction(TestSuite suite, String functionName, ClassLoader classLoader,
                                    Scheduler scheduler, Class<?>[] types, Object[] args) throws ClassNotFoundException {
         Class<?> functionClass = classLoader.loadClass(suite.getTestUtilityFunctions().get(functionName));
         TesterinaFunction testerinaFunction = new TesterinaFunction(functionClass, functionName, scheduler);
-        testerinaFunction.invoke(types, args);
+        return testerinaFunction.invoke(types, args);
     }
 
     private String formatErrorMessage(Throwable e) {
         String message;
-        if (e.getCause() instanceof BError) {
-            try {
-                message = ((BError) e.getCause()).getPrintableStackTrace();
-            } catch (ClassCastException castException) {
-                // throw the exception to top
+
+        try {
+            if (e instanceof BError) {
+                message = ((BError) e).getPrintableStackTrace();
+            } else if (e instanceof Exception | e instanceof Error) {
+                RuntimeUtils.silentlyLogBadSad(e);
+                message = TesterinaUtils.getPrintableStackTrace(e);
+            } else {
                 throw new BallerinaTestException(e);
             }
-        } else if (e instanceof BallerinaTestException) {
-            try {
-                message = TesterinaUtils.getPrintableStackTrace(e.getCause());
-                RuntimeUtils.silentlyLogBadSad(e.getCause());
-            } catch (ClassCastException castException) {
-                throw new BallerinaTestException(e);
-            }
-        } else {
-            throw new BallerinaTestException(e);
+        } catch (ClassCastException classCastException) {
+            // If an unhandled error type is passed to format error message
+            throw new BallerinaTestException(classCastException);
         }
+
         return message;
     }
 

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java
@@ -672,11 +672,13 @@ public class BTestRunner {
                 RuntimeUtils.silentlyLogBadSad(e);
                 return TesterinaUtils.getPrintableStackTrace(e);
             } else {
-                throw new BallerinaTestException(e);
+                RuntimeUtils.silentlyLogBadSad(e);
+                return TesterinaUtils.getPrintableStackTrace(e);
             }
         } catch (ClassCastException classCastException) {
             // If an unhandled error type is passed to format error message
-            throw new BallerinaTestException(classCastException);
+            RuntimeUtils.silentlyLogBadSad(e);
+            return TesterinaUtils.getPrintableStackTrace(e);
         }
     }
 

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java
@@ -222,9 +222,9 @@ public class BTestRunner {
         ClassLoader classLoader = testClassLoader;
         // Load module init class
         String initClassName = TesterinaUtils.getQualifiedClassName(suite.getOrgName(),
-                                                                    suite.getPackageID(),
-                                                                    suite.getVersion(),
-                                                                    MODULE_INIT_CLASS_NAME);
+                suite.getPackageID(),
+                suite.getVersion(),
+                MODULE_INIT_CLASS_NAME);
         Class<?> initClazz;
         try {
             initClazz = classLoader.loadClass(initClassName);
@@ -287,14 +287,14 @@ public class BTestRunner {
                 new Object[]{new String[]{}, getConfigPaths(suite), null, null});
         if (response instanceof Throwable) {
             throw new BallerinaTestException("Configurable initialization for test suite failed due to " +
-                    response.toString(), (Throwable) response);
+                    formatErrorMessage((Throwable) response), (Throwable) response);
         }
 
         init.setName("$moduleInit");
         response = init.invoke();
         if (response instanceof Throwable) {
             throw new BallerinaTestException("Dependant module initialization for test suite failed due to " +
-                    response.toString(), (Throwable) response);
+                    formatErrorMessage((Throwable) response), (Throwable) response);
         }
         // Now we initialize the init of testable module.
         if (hasTestablePackage) {
@@ -303,7 +303,7 @@ public class BTestRunner {
             response = testInit.invoke();
             if (response instanceof Throwable) {
                 throw new BallerinaTestException("Test module initialization for test suite failed due to " +
-                        response.toString(), (Throwable) response);
+                        formatErrorMessage((Throwable) response), (Throwable) response);
             }
         }
         // As the start function we need to use $moduleStart to start all the dependent modules
@@ -377,7 +377,7 @@ public class BTestRunner {
             executeBeforeFunction(test, suite, classLoader, scheduler, shouldSkip, shouldSkipTest);
             // run the test
             executeFunction(test, suite, packageName, classLoader, scheduler, shouldSkip, shouldSkipTest,
-                            failedOrSkippedTests, failedAfterFuncTests);
+                    failedOrSkippedTests, failedAfterFuncTests);
             // run the after tests
             executeAfterFunction(test, suite, classLoader, scheduler, shouldSkip, shouldSkipTest, failedAfterFuncTests);
             // run the after each tests
@@ -390,7 +390,7 @@ public class BTestRunner {
     }
 
     private void executeBeforeGroupFunctions(Test test, TestSuite suite, ClassLoader classLoader, Scheduler scheduler,
-                                       AtomicBoolean shouldSkip, AtomicBoolean shouldSkipTest,
+                                             AtomicBoolean shouldSkip, AtomicBoolean shouldSkipTest,
                                              AtomicBoolean shouldSkipAfterGroups)  {
         if (!shouldSkip.get() && !shouldSkipTest.get()) {
             for (String groupName : test.getGroups()) {
@@ -434,9 +434,9 @@ public class BTestRunner {
                 } catch (Throwable e) {
                     shouldSkip.set(true);
                     errorMsg = String.format("\t[fail] " + beforeEachTest +
-                                                     " [before each test function for the test %s] :\n\t    %s",
-                                             test,
-                                             formatErrorMessage(e));
+                                    " [before each test function for the test %s] :\n\t    %s",
+                            test,
+                            formatErrorMessage(e));
                     errStream.println(errorMsg);
                 }
             });
@@ -458,8 +458,8 @@ public class BTestRunner {
             } catch (Throwable e) {
                 shouldSkipTest.set(true);
                 errorMsg = String.format("\t[fail] " + test.getBeforeTestFunction() +
-                                                 " [before test function for the test %s] :\n\t    %s",
-                                         test, formatErrorMessage(e));
+                                " [before test function for the test %s] :\n\t    %s",
+                        test, formatErrorMessage(e));
                 errStream.println(errorMsg);
             }
         }
@@ -470,43 +470,34 @@ public class BTestRunner {
                                  List<String> failedOrSkippedTests, List<String> failedAfterFuncTests) {
         TesterinaResult functionResult;
 
-        try {
-            if (isTestDependsOnFailedFunctions(test.getDependsOnTestFunctions(), failedOrSkippedTests) ||
+        if (isTestDependsOnFailedFunctions(test.getDependsOnTestFunctions(), failedOrSkippedTests) ||
                 isTestDependsOnFailedFunctions(test.getDependsOnTestFunctions(), failedAfterFuncTests)) {
-                shouldSkipTest.set(true);
-            }
+            shouldSkipTest.set(true);
+        }
 
-            // Check whether the this test depends on any failed or skipped functions
-            if (!shouldSkip.get() && !shouldSkipTest.get()) {
-                Object valueSets = null;
-                if (test.getDataProvider() != null) {
-                    valueSets = invokeTestFunction(suite, test.getDataProvider(), classLoader, scheduler);
-                }
-                if (valueSets == null) {
-                    valueSets = invokeTestFunction(suite, test.getTestName(), classLoader, scheduler);
-                    computeFunctionResult(test, packageName, shouldSkip, failedOrSkippedTests, valueSets);
-                } else {
-                    Class<?>[] argTypes = extractArgumentTypes(valueSets);
-                    List<Object[]> argList = extractArguments(valueSets);
-                    for (Object[] arg : argList) {
-                        valueSets = invokeTestFunction(suite, test.getTestName(), classLoader, scheduler, argTypes,
-                                arg);
-                        computeFunctionResult(test, packageName, shouldSkip, failedOrSkippedTests, valueSets);
-                    }
-                }
-            } else {
-                // If the test function is skipped lets add it to the failed test list
-                failedOrSkippedTests.add(test.getTestName());
-                // report the test result
-                functionResult = new TesterinaResult(test.getTestName(), false, true, null);
-                tReport.addFunctionResult(packageName, functionResult);
+        // Check whether the this test depends on any failed or skipped functions
+        if (!shouldSkip.get() && !shouldSkipTest.get()) {
+            Object valueSets = null;
+            if (test.getDataProvider() != null) {
+                valueSets = invokeTestFunction(suite, test.getDataProvider(), classLoader, scheduler);
             }
-        } catch (BallerinaTestException | ClassNotFoundException e) {
+            if (valueSets == null) {
+                valueSets = invokeTestFunction(suite, test.getTestName(), classLoader, scheduler);
+                computeFunctionResult(test, packageName, shouldSkip, failedOrSkippedTests, valueSets);
+            } else {
+                Class<?>[] argTypes = extractArgumentTypes(valueSets);
+                List<Object[]> argList = extractArguments(valueSets);
+                for (Object[] arg : argList) {
+                    valueSets = invokeTestFunction(suite, test.getTestName(), classLoader, scheduler, argTypes,
+                            arg);
+                    computeFunctionResult(test, packageName, shouldSkip, failedOrSkippedTests, valueSets);
+                }
+            }
+        } else {
             // If the test function is skipped lets add it to the failed test list
             failedOrSkippedTests.add(test.getTestName());
             // report the test result
-            functionResult = new TesterinaResult(test.getTestName(), false, shouldSkip.get(),
-                                                 formatErrorMessage(e));
+            functionResult = new TesterinaResult(test.getTestName(), false, true, null);
             tReport.addFunctionResult(packageName, functionResult);
         }
         for (String groupName : test.getGroups()) {
@@ -587,7 +578,7 @@ public class BTestRunner {
     }
 
     private void executeAfterGroupFunctions(Test test, TestSuite suite, ClassLoader classLoader, Scheduler scheduler,
-                                             AtomicBoolean shouldSkip, AtomicBoolean shouldSkipTest,
+                                            AtomicBoolean shouldSkip, AtomicBoolean shouldSkipTest,
                                             AtomicBoolean shouldSkipAfterGroups)  {
         if (!shouldSkipAfterGroups.get() && !shouldSkip.get() && !shouldSkipTest.get()) {
             for (String groupName : test.getGroups()) {
@@ -652,27 +643,34 @@ public class BTestRunner {
     }
 
     private Object invokeTestFunction(TestSuite suite, String functionName, ClassLoader classLoader,
-                                      Scheduler scheduler) throws ClassNotFoundException {
-        Class<?> functionClass = classLoader.loadClass(suite.getTestUtilityFunctions().get(functionName));
-        TesterinaFunction testerinaFunction = new TesterinaFunction(functionClass, functionName, scheduler);
-        return testerinaFunction.invoke();
+                                      Scheduler scheduler) {
+        try {
+            Class<?> functionClass = classLoader.loadClass(suite.getTestUtilityFunctions().get(functionName));
+            TesterinaFunction testerinaFunction = new TesterinaFunction(functionClass, functionName, scheduler);
+            return testerinaFunction.invoke();
+        } catch (ClassNotFoundException e) {
+            return e;
+        }
     }
 
     public Object invokeTestFunction(TestSuite suite, String functionName, ClassLoader classLoader,
-                                   Scheduler scheduler, Class<?>[] types, Object[] args) throws ClassNotFoundException {
-        Class<?> functionClass = classLoader.loadClass(suite.getTestUtilityFunctions().get(functionName));
-        TesterinaFunction testerinaFunction = new TesterinaFunction(functionClass, functionName, scheduler);
-        return testerinaFunction.invoke(types, args);
+                                     Scheduler scheduler, Class<?>[] types, Object[] args) {
+        try {
+            Class<?> functionClass = classLoader.loadClass(suite.getTestUtilityFunctions().get(functionName));
+            TesterinaFunction testerinaFunction = new TesterinaFunction(functionClass, functionName, scheduler);
+            return testerinaFunction.invoke(types, args);
+        } catch (ClassNotFoundException e) {
+            return e;
+        }
     }
 
     private String formatErrorMessage(Throwable e) {
-        String message;
         try {
             if (e instanceof BError) {
-                message = ((BError) e).getPrintableStackTrace();
+                return ((BError) e).getPrintableStackTrace();
             } else if (e instanceof Exception | e instanceof Error) {
                 RuntimeUtils.silentlyLogBadSad(e);
-                message = TesterinaUtils.getPrintableStackTrace(e);
+                return TesterinaUtils.getPrintableStackTrace(e);
             } else {
                 throw new BallerinaTestException(e);
             }
@@ -680,7 +678,6 @@ public class BTestRunner {
             // If an unhandled error type is passed to format error message
             throw new BallerinaTestException(classCastException);
         }
-        return message;
     }
 
     private boolean isTestDependsOnFailedFunctions(List<String> failedOrSkippedTests, List<String> dependentTests) {

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java
@@ -270,9 +270,75 @@ public class BTestRunner {
         // Run After suite functions
         executeAfterSuiteFunctions(suite, classLoader, scheduler, shouldSkipAfterSuite);
         // Call module stop and test stop function
-        stopSuite(suite, scheduler, initClazz, testInitClazz, hasTestablePackage);
+        stopSuite(scheduler, initClazz, testInitClazz, hasTestablePackage);
         // print module test results
         tReport.printTestSuiteSummary(packageName);
+    }
+
+    private void startSuite(TestSuite suite, Scheduler initScheduler, Class<?> initClazz, Class<?> testInitClazz,
+                            Class<?> configClazz, boolean hasTestablePackage) {
+        TesterinaFunction init = new TesterinaFunction(initClazz, INIT_FUNCTION_NAME, initScheduler);
+        TesterinaFunction start = new TesterinaFunction(initClazz, START_FUNCTION_NAME, initScheduler);
+        TesterinaFunction configInit = new TesterinaFunction(configClazz, "$configureInit", initScheduler);
+        // As the init function we need to use $moduleInit to initialize all the dependent modules
+        // properly.
+
+        Object response = configInit.directInvoke(new Class[]{String[].class, Path[].class, String.class, String.class},
+                new Object[]{new String[]{}, getConfigPaths(suite), null, null});
+        if (response instanceof Throwable) {
+            throw new BallerinaTestException("Configurable initialization for test suite failed due to " +
+                    response.toString(), (Throwable) response);
+        }
+
+        init.setName("$moduleInit");
+        response = init.invoke();
+        if (response instanceof Throwable) {
+            throw new BallerinaTestException("Dependant module initialization for test suite failed due to " +
+                    response.toString(), (Throwable) response);
+        }
+        // Now we initialize the init of testable module.
+        if (hasTestablePackage) {
+            TesterinaFunction testInit =
+                    new TesterinaFunction(testInitClazz, TEST_INIT_FUNCTION_NAME, initScheduler);
+            response = testInit.invoke();
+            if (response instanceof Throwable) {
+                throw new BallerinaTestException("Test module initialization for test suite failed due to " +
+                        response.toString(), (Throwable) response);
+            }
+        }
+        // As the start function we need to use $moduleStart to start all the dependent modules
+        // properly.
+        start.setName("$moduleStart");
+        start.invoke();
+        // Invoke start function of the testable module
+        if (hasTestablePackage) {
+            TesterinaFunction testStart =
+                    new TesterinaFunction(testInitClazz, TEST_START_FUNCTION_NAME, initScheduler);
+            response = testStart.invoke();
+            if (response instanceof BError || response instanceof Throwable || response instanceof Error) {
+                throw new BallerinaTestException("Test module invocation for test suite failed due to " +
+                        response.toString(), (Throwable) response);
+            }
+        }
+        // Once the start function finish we will re start the scheduler with immortal true
+        initScheduler.setImmortal(true);
+        Thread immortalThread = new Thread(initScheduler::start, "module-start");
+        immortalThread.setDaemon(true);
+        immortalThread.start();
+    }
+
+    private Path[] getConfigPaths(TestSuite testSuite) {
+        String moduleName = testSuite.getModuleName();
+        Path configFilePath = Paths.get(testSuite.getSourceRootPath());
+        if (!moduleName.equals(testSuite.getPackageName())) {
+            configFilePath = configFilePath.resolve(ProjectConstants.MODULES_ROOT).resolve(moduleName);
+        }
+        configFilePath = configFilePath.resolve(ProjectConstants.TEST_DIR_NAME).resolve(CONFIG_FILE_NAME);
+        if (!Files.exists(configFilePath)) {
+            return new Path[] {};
+        } else {
+            return new Path[] {configFilePath};
+        }
     }
 
     private void executeBeforeSuiteFunctions(TestSuite suite, ClassLoader classLoader, Scheduler scheduler,
@@ -280,7 +346,10 @@ public class BTestRunner {
         suite.getBeforeSuiteFunctionNames().forEach(test -> {
             String errorMsg;
             try {
-                invokeTestFunction(suite, test, classLoader, scheduler);
+                Object value = invokeTestFunction(suite, test, classLoader, scheduler);
+                if (value instanceof BError || value instanceof  Exception || value instanceof Error) {
+                    throw (Throwable) value;
+                }
             } catch (Throwable e) {
                 shouldSkip.set(true);
                 shouldSkipAfterSuite.set(true);
@@ -331,7 +400,10 @@ public class BTestRunner {
                     String errorMsg;
                     for (String beforeGroupFunc : suite.getGroups().get(groupName).getBeforeGroupsFunctions()) {
                         try {
-                            invokeTestFunction(suite, beforeGroupFunc, classLoader, scheduler);
+                            Object value = invokeTestFunction(suite, beforeGroupFunc, classLoader, scheduler);
+                            if (value instanceof BError || value instanceof  Exception || value instanceof Error) {
+                                throw (Throwable) value;
+                            }
                         } catch (Throwable e) {
                             shouldSkip.set(true);
                             shouldSkipTest.set(true);
@@ -355,7 +427,10 @@ public class BTestRunner {
             suite.getBeforeEachFunctionNames().forEach(beforeEachTest -> {
                 String errorMsg;
                 try {
-                    invokeTestFunction(suite, beforeEachTest, classLoader, scheduler);
+                    Object value = invokeTestFunction(suite, beforeEachTest, classLoader, scheduler);
+                    if (value instanceof BError || value instanceof  Exception || value instanceof Error) {
+                        throw (Throwable) value;
+                    }
                 } catch (Throwable e) {
                     shouldSkip.set(true);
                     errorMsg = String.format("\t[fail] " + beforeEachTest +
@@ -375,7 +450,10 @@ public class BTestRunner {
             String errorMsg;
             try {
                 if (test.getBeforeTestFunction() != null) {
-                    invokeTestFunction(suite, test.getBeforeTestFunction(), classLoader, scheduler);
+                    Object value = invokeTestFunction(suite, test.getBeforeTestFunction(), classLoader, scheduler);
+                    if (value instanceof BError || value instanceof  Exception || value instanceof Error) {
+                        throw (Throwable) value;
+                    }
                 }
             } catch (Throwable e) {
                 shouldSkipTest.set(true);
@@ -474,7 +552,10 @@ public class BTestRunner {
         if (!shouldSkip.get() && !shouldSkipTest.get()) {
             try {
                 if (test.getAfterTestFunction() != null) {
-                    invokeTestFunction(suite, test.getAfterTestFunction(), classLoader, scheduler);
+                    Object value = invokeTestFunction(suite, test.getAfterTestFunction(), classLoader, scheduler);
+                    if (value instanceof BError || value instanceof  Exception || value instanceof Error) {
+                        throw (Throwable) value;
+                    }
                 }
             } catch (Throwable e) {
                 failedAfterFuncTests.add(test.getTestName());
@@ -490,7 +571,10 @@ public class BTestRunner {
         if (!shouldSkip.get() && !shouldSkipTest.get()) {
             suite.getAfterEachFunctionNames().forEach(afterEachTest -> {
                 try {
-                    invokeTestFunction(suite, afterEachTest, classLoader, scheduler);
+                    Object value = invokeTestFunction(suite, afterEachTest, classLoader, scheduler);
+                    if (value instanceof BError || value instanceof  Exception || value instanceof Error) {
+                        throw (Throwable) value;
+                    }
                 } catch (Throwable e) {
                     shouldSkip.set(true);
                     String errorMsg = String.format("\t[fail] " + afterEachTest +
@@ -513,7 +597,10 @@ public class BTestRunner {
                     String errorMsg;
                     for (String afterGroupFunc : suite.getGroups().get(groupName).getAfterGroupsFunctions()) {
                         try {
-                            invokeTestFunction(suite, afterGroupFunc, classLoader, scheduler);
+                            Object value = invokeTestFunction(suite, afterGroupFunc, classLoader, scheduler);
+                            if (value instanceof BError || value instanceof  Exception || value instanceof Error) {
+                                throw (Throwable) value;
+                            }
                         } catch (Throwable e) {
                             shouldSkip.set(true);
                             shouldSkipTest.set(true);
@@ -536,7 +623,10 @@ public class BTestRunner {
             if (!shouldSkipAfterSuite.get() || alwaysRun.get()) {
                 String errorMsg;
                 try {
-                    invokeTestFunction(suite, func, classLoader, scheduler);
+                    Object value = invokeTestFunction(suite, func, classLoader, scheduler);
+                    if (value instanceof BError || value instanceof  Exception || value instanceof Error) {
+                        throw (Throwable) value;
+                    }
                 } catch (Throwable e) {
                     errorMsg = String.format("\t[fail] " + func + " [after test suite function] :\n\t    " +
                             "%s", formatErrorMessage(e));
@@ -546,69 +636,7 @@ public class BTestRunner {
         });
     }
 
-    private void startSuite(TestSuite suite, Scheduler initScheduler, Class<?> initClazz, Class<?> testInitClazz,
-                            Class<?> configClazz, boolean hasTestablePackage) {
-        TesterinaFunction init = new TesterinaFunction(initClazz, INIT_FUNCTION_NAME, initScheduler);
-        TesterinaFunction start = new TesterinaFunction(initClazz, START_FUNCTION_NAME, initScheduler);
-        TesterinaFunction configInit = new TesterinaFunction(configClazz, "$configureInit", initScheduler);
-        // As the init function we need to use $moduleInit to initialize all the dependent modules
-        // properly.
-
-        Object response = configInit.directInvoke(new Class[]{String[].class, Path[].class, String.class, String.class},
-                new Object[]{new String[]{}, getConfigPaths(suite), null, null});
-        if (response instanceof Throwable) {
-            throw new BallerinaTestException("Configurable initialization for test suite failed due to " +
-                    response.toString(), (Throwable) response);
-        }
-
-        init.setName("$moduleInit");
-        response = init.invoke();
-        if (response instanceof Throwable) {
-            throw new BallerinaTestException("Dependant module initialization for test suite failed due to " +
-                    response.toString(), (Throwable) response);
-        }
-        // Now we initialize the init of testable module.
-        if (hasTestablePackage) {
-            TesterinaFunction testInit =
-                    new TesterinaFunction(testInitClazz, TEST_INIT_FUNCTION_NAME, initScheduler);
-            response = testInit.invoke();
-            if (response instanceof Throwable) {
-                throw new BallerinaTestException("Test module initialization for test suite failed due to " +
-                        response.toString(), (Throwable) response);
-            }
-        }
-        // As the start function we need to use $moduleStart to start all the dependent modules
-        // properly.
-        start.setName("$moduleStart");
-        start.invoke();
-        // Invoke start function of the testable module
-        if (hasTestablePackage) {
-            TesterinaFunction testStart =
-                    new TesterinaFunction(testInitClazz, TEST_START_FUNCTION_NAME, initScheduler);
-            testStart.invoke();
-        }
-        // Once the start function finish we will re start the scheduler with immortal true
-        initScheduler.setImmortal(true);
-        Thread immortalThread = new Thread(initScheduler::start, "module-start");
-        immortalThread.setDaemon(true);
-        immortalThread.start();
-    }
-
-    private Path[] getConfigPaths(TestSuite testSuite) {
-        String moduleName = testSuite.getModuleName();
-        Path configFilePath = Paths.get(testSuite.getSourceRootPath());
-        if (!moduleName.equals(testSuite.getPackageName())) {
-            configFilePath = configFilePath.resolve(ProjectConstants.MODULES_ROOT).resolve(moduleName);
-        }
-        configFilePath = configFilePath.resolve(ProjectConstants.TEST_DIR_NAME).resolve(CONFIG_FILE_NAME);
-        if (!Files.exists(configFilePath)) {
-            return new Path[] {};
-        } else {
-            return new Path[] {configFilePath};
-        }
-    }
-
-    private void stopSuite(TestSuite suite, Scheduler scheduler, Class<?> initClazz, Class<?> testInitClazz,
+    private void stopSuite(Scheduler scheduler, Class<?> initClazz, Class<?> testInitClazz,
                            boolean hasTestablePackage) {
         TesterinaFunction stop = new TesterinaFunction(initClazz, STOP_FUNCTION_NAME, scheduler);
         // Invoke stop function of the testable module.
@@ -639,7 +667,6 @@ public class BTestRunner {
 
     private String formatErrorMessage(Throwable e) {
         String message;
-
         try {
             if (e instanceof BError) {
                 message = ((BError) e).getPrintableStackTrace();
@@ -653,7 +680,6 @@ public class BTestRunner {
             // If an unhandled error type is passed to format error message
             throw new BallerinaTestException(classCastException);
         }
-
         return message;
     }
 

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/TesterinaFunction.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/TesterinaFunction.java
@@ -19,7 +19,6 @@ package org.ballerinalang.test.runtime.entity;
 
 import io.ballerina.runtime.api.PredefinedTypes;
 import io.ballerina.runtime.api.utils.IdentifierUtils;
-import io.ballerina.runtime.api.values.BError;
 import io.ballerina.runtime.api.values.BFuture;
 import io.ballerina.runtime.internal.scheduling.Scheduler;
 import io.ballerina.runtime.internal.scheduling.Strand;
@@ -118,12 +117,6 @@ public class TesterinaFunction {
             scheduler.start();
             final Throwable t = out.getPanic();
             final Object result = out.getResult();
-            if (result instanceof BError) {
-                throw new BallerinaTestException((BError) result);
-            }
-            if (result instanceof Exception) {
-                throw new BallerinaTestException((Exception) result);
-            }
             if (t != null) {
                 throw new BallerinaTestException("Error while invoking function '" + funcName + "'", t.getMessage());
             }

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/TesterinaFunction.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/TesterinaFunction.java
@@ -109,16 +109,15 @@ public class TesterinaFunction {
                 } catch (InvocationTargetException e) {
                     return e.getTargetException();
                 } catch (IllegalAccessException e) {
-                    throw new BallerinaTestException("Error while invoking function '" + funcName + "'", e);
+                    return new BallerinaTestException("Error while invoking function '" + funcName + "'", e);
                 }
             };
             final BFuture out = scheduler.schedule(params, func, null, null, null, PredefinedTypes.TYPE_ANY,
                                                    null, null);
             scheduler.start();
             final Throwable t = out.getPanic();
-            final Object result = out.getResult();
             if (t != null) {
-                throw new BallerinaTestException("Error while invoking function '" + funcName + "'", t.getMessage());
+                return new BallerinaTestException("Error while invoking function '" + funcName + "'", t.getMessage());
             }
             return out.getResult();
         } catch (NoSuchMethodException e) {

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/TesterinaFunction.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/TesterinaFunction.java
@@ -122,7 +122,7 @@ public class TesterinaFunction {
             }
             return out.getResult();
         } catch (NoSuchMethodException e) {
-            throw new BallerinaTestException("Error while invoking function '" + funcName + "'\n" +
+            return new BallerinaTestException("Error while invoking function '" + funcName + "'\n" +
                     "If you are using data providers please check if types return from data provider " +
                     "match test function parameter types.", e);
         }

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/TesterinaUtils.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/TesterinaUtils.java
@@ -92,7 +92,7 @@ public class TesterinaUtils {
                 throw new RuntimeException("there are test failures");
             }
         } catch (BallerinaTestException e) {
-            errStream.println("Runtime error: " + e);
+            errStream.println("error: " + e);
             RuntimeUtils.silentlyLogBadSad(e);
             throw e;
         } catch (Throwable e) {
@@ -296,6 +296,7 @@ public class TesterinaUtils {
         for (int i = 1; i < stackTrace.length; i++) {
             printStackElement(sb, stackTrace[i], "\n\t   ");
         }
+        sb.append("\n\nCheck ballerina-internal log for more details");
         return sb.toString();
 
     }

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/TesterinaUtils.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/TesterinaUtils.java
@@ -17,6 +17,7 @@
  */
 package org.ballerinalang.test.runtime.util;
 
+import io.ballerina.runtime.api.utils.IdentifierUtils;
 import io.ballerina.runtime.internal.util.RuntimeUtils;
 import org.ballerinalang.test.runtime.BTestRunner;
 import org.ballerinalang.test.runtime.entity.Test;
@@ -32,8 +33,12 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.Comparator;
+import java.util.LinkedList;
 import java.util.List;
+import java.util.Optional;
 
+import static io.ballerina.runtime.api.constants.RuntimeConstants.BLANG_SRC_FILE_SUFFIX;
+import static io.ballerina.runtime.api.constants.RuntimeConstants.MODULE_INIT_CLASS_NAME;
 import static io.ballerina.runtime.api.utils.IdentifierUtils.encodeNonFunctionIdentifier;
 import static org.ballerinalang.test.runtime.util.TesterinaConstants.ANON_ORG;
 import static org.ballerinalang.test.runtime.util.TesterinaConstants.DOT;
@@ -45,6 +50,14 @@ public class TesterinaUtils {
 
     private static final PrintStream outStream = System.out;
     private static final PrintStream errStream = System.err;
+
+    private static final String GENERATE_OBJECT_CLASS_PREFIX = ".$value$";
+    private static final String GENERATE_PKG_INIT = "___init_";
+    private static final String GENERATE_PKG_START = "___start_";
+    private static final String GENERATE_PKG_STOP = "___stop_";
+    private static final String INIT_FUNCTION_SUFFIX = "..<init>";
+    private static final String START_FUNCTION_SUFFIX = ".<start>";
+    private static final String STOP_FUNCTION_SUFFIX = ".<stop>";
 
     /**
      * Cleans up any remaining testerina metadata.
@@ -79,11 +92,10 @@ public class TesterinaUtils {
                 throw new RuntimeException("there are test failures");
             }
         } catch (BallerinaTestException e) {
-            errStream.println("error: " + e.getCause());
-            RuntimeUtils.silentlyLogBadSad(e.getCause());
+            errStream.println("Runtime error: " + e);
+            RuntimeUtils.silentlyLogBadSad(e);
             throw e;
         } catch (Throwable e) {
-            RuntimeUtils.silentlyLogBadSad(e);
             throw new RuntimeException("test execution failed due to runtime exception");
         }
     }
@@ -267,6 +279,105 @@ public class TesterinaUtils {
     public static List<org.ballerinalang.test.runtime.entity.Test> getSingleExecutionTestsOld(
             List<org.ballerinalang.test.runtime.entity.Test> currentTests, List<String> functions) {
         return Collections.emptyList();
+    }
+
+    public static String getPrintableStackTrace(Throwable throwable) {
+        String errorMsg = throwable.toString();
+        StringBuilder sb = new StringBuilder();
+        sb.append(errorMsg);
+        // Append function/action/resource name with package path (if any)
+        StackTraceElement[] stackTrace = getStackTrace(throwable);
+        if (stackTrace.length == 0) {
+            return sb.toString();
+        }
+        sb.append("\n\tat ");
+        // print first element
+        printStackElement(sb, stackTrace[0], "");
+        for (int i = 1; i < stackTrace.length; i++) {
+            printStackElement(sb, stackTrace[i], "\n\t   ");
+        }
+        return sb.toString();
+
+    }
+
+    private static void printStackElement(StringBuilder sb, StackTraceElement stackTraceElement, String tab) {
+        String pkgName = IdentifierUtils.decodeIdentifier(stackTraceElement.getClassName());
+        String fileName = stackTraceElement.getFileName();
+
+        // clean file name from pkgName since we print the file name after the method name.
+        fileName = fileName.replace(BLANG_SRC_FILE_SUFFIX, "");
+        fileName = fileName.replace("/", "-");
+        int index = pkgName.lastIndexOf("." + fileName);
+        if (index != -1) {
+            pkgName = pkgName.substring(0, index);
+        }
+        // todo we need to seperate orgname and module name with '/'
+
+        sb.append(tab);
+        if (!pkgName.equals(MODULE_INIT_CLASS_NAME)) {
+            sb.append(pkgName).append(":");
+        }
+
+        // Append the method name
+        sb.append(IdentifierUtils.decodeIdentifier(stackTraceElement.getMethodName()));
+        // Append the filename
+        sb.append("(").append(stackTraceElement.getFileName());
+        // Append the line number
+        sb.append(":").append(stackTraceElement.getLineNumber()).append(")");
+    }
+
+    public static StackTraceElement[] getStackTrace(Throwable throwable) {
+        StackTraceElement[] stackTrace = throwable.getStackTrace();
+        List<StackTraceElement> filteredStack = new LinkedList<>();
+        int index = 0;
+        for (StackTraceElement stackFrame : stackTrace) {
+            Optional<StackTraceElement> stackTraceElement = filterStackTraceElement(stackFrame, index++);
+            stackTraceElement.ifPresent(filteredStack::add);
+        }
+        StackTraceElement[] filteredStackArray = new StackTraceElement[filteredStack.size()];
+        return filteredStack.toArray(filteredStackArray);
+    }
+
+    private static Optional<StackTraceElement> filterStackTraceElement(StackTraceElement stackFrame, int currentIndex) {
+        String fileName = stackFrame.getFileName();
+        int lineNo = stackFrame.getLineNumber();
+        if (lineNo < 0) {
+            return Optional.empty();
+        }
+        // Handle init function
+        String className = stackFrame.getClassName();
+        String methodName = stackFrame.getMethodName();
+        if (className.equals(MODULE_INIT_CLASS_NAME)) {
+            if (currentIndex == 0) {
+                return Optional.empty();
+            }
+            switch (methodName) {
+                case GENERATE_PKG_INIT:
+                    methodName = INIT_FUNCTION_SUFFIX;
+                    break;
+                case GENERATE_PKG_START:
+                    methodName = START_FUNCTION_SUFFIX;
+                    break;
+                case GENERATE_PKG_STOP:
+                    methodName = STOP_FUNCTION_SUFFIX;
+                    break;
+                default:
+                    return Optional.empty();
+            }
+            return Optional.of(new StackTraceElement(cleanupClassName(className), methodName, fileName,
+                    stackFrame.getLineNumber()));
+
+        }
+        if (fileName != null && !fileName.endsWith(BLANG_SRC_FILE_SUFFIX)) {
+            // Remove java sources for bal stacktrace if they are not extern functions.
+            return Optional.empty();
+        }
+        return Optional.of(
+                new StackTraceElement(cleanupClassName(className), methodName, fileName, stackFrame.getLineNumber()));
+    }
+
+    private static String cleanupClassName(String className) {
+        return className.replace(GENERATE_OBJECT_CLASS_PREFIX, ".");
     }
 
 }

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/TesterinaUtils.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/TesterinaUtils.java
@@ -92,7 +92,7 @@ public class TesterinaUtils {
                 throw new RuntimeException("there are test failures");
             }
         } catch (BallerinaTestException e) {
-            errStream.println("error: " + e);
+            errStream.println("error: " + e.getMessage());
             RuntimeUtils.silentlyLogBadSad(e);
             throw e;
         } catch (Throwable e) {

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/TesterinaUtils.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/TesterinaUtils.java
@@ -18,7 +18,6 @@
 package org.ballerinalang.test.runtime.util;
 
 import io.ballerina.runtime.api.utils.IdentifierUtils;
-import io.ballerina.runtime.internal.util.RuntimeUtils;
 import org.ballerinalang.test.runtime.BTestRunner;
 import org.ballerinalang.test.runtime.entity.Test;
 import org.ballerinalang.test.runtime.entity.TestSuite;
@@ -93,7 +92,6 @@ public class TesterinaUtils {
             }
         } catch (BallerinaTestException e) {
             errStream.println("error: " + e.getMessage());
-            RuntimeUtils.silentlyLogBadSad(e);
             throw e;
         } catch (Throwable e) {
             throw new RuntimeException("test execution failed due to runtime exception");
@@ -298,7 +296,6 @@ public class TesterinaUtils {
         }
         sb.append("\n\nCheck ballerina-internal log for more details");
         return sb.toString();
-
     }
 
     private static void printStackElement(StringBuilder sb, StackTraceElement stackTraceElement, String tab) {

--- a/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/TesterinaUtils.java
+++ b/misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/util/TesterinaUtils.java
@@ -79,8 +79,8 @@ public class TesterinaUtils {
                 throw new RuntimeException("there are test failures");
             }
         } catch (BallerinaTestException e) {
-            errStream.println("error: " + e.getMessage());
-            RuntimeUtils.silentlyLogBadSad(e);
+            errStream.println("error: " + e.getCause());
+            RuntimeUtils.silentlyLogBadSad(e.getCause());
             throw e;
         } catch (Throwable e) {
             RuntimeUtils.silentlyLogBadSad(e);

--- a/tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/negative/InvalidDataProviderTestCase.java
+++ b/tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/negative/InvalidDataProviderTestCase.java
@@ -54,39 +54,45 @@ public class InvalidDataProviderTestCase extends BaseTestCase {
 
     @Test
     public void testInvalidDataProvider() throws BallerinaTestException {
-        String errMsg = "error: Error while invoking function 'testInvalidDataProvider'";
-        String errMsg2 = "If you are using data providers please check if types return from data provider match test " +
+        String errMsg1 = "[fail] testInvalidDataProvider:";
+        String errMsg2 = "org.ballerinalang.test.runtime.exceptions.BallerinaTestException: " +
+                "Error while invoking function 'testInvalidDataProvider'";
+        String errMsg3 = "If you are using data providers please check if types return from data provider match test " +
                 "function parameter types";
         String[] args = mergeCoverageArgs(new String[]{"invalid-data-provider-test.bal"});
         String output = balClient.runMainAndReadStdOut("test", args,
-                new HashMap<>(), projectPath, true);
-        if (!output.contains(errMsg) || !output.contains(errMsg2)) {
+                new HashMap<>(), projectPath, false);
+        if (!output.contains(errMsg1) || !output.contains(errMsg2) || !output.contains(errMsg3)) {
             AssertionUtils.assertForTestFailures(output, "error while handling invalid data providers");
         }
     }
 
     @Test
     public void testInvalidDataProvider2() throws BallerinaTestException {
-        String errMsg = "error: Error while invoking function 'testInvalidDataProvider2'";
-        String errMsg2 = "If you are using data providers please check if types return from data provider match test " +
+        String errMsg1 = "[fail] testInvalidDataProvider2:";
+        String errMsg2 = "org.ballerinalang.test.runtime.exceptions.BallerinaTestException: " +
+                "Error while invoking function 'testInvalidDataProvider2'";
+        String errMsg3 = "If you are using data providers please check if types return from data provider match test " +
                 "function parameter types";
         String[] args = mergeCoverageArgs(new String[]{"invalid-data-provider-test2.bal"});
         String output = balClient.runMainAndReadStdOut("test", args,
-                new HashMap<>(), projectPath, true);
-        if (!output.contains(errMsg) || !output.contains(errMsg2)) {
+                new HashMap<>(), projectPath, false);
+        if (!output.contains(errMsg1) || !output.contains(errMsg2) || !output.contains(errMsg3)) {
             AssertionUtils.assertForTestFailures(output, "error while handling invalid data providers");
         }
     }
 
     @Test
     public void testInvalidTupleDataProvider() throws BallerinaTestException {
-        String errMsg = "error: Error while invoking function 'testInvalidTupleDataProvider'";
-        String errMsg2 = "If you are using data providers please check if types return from data provider match" +
+        String errMsg1 = "[fail] testInvalidTupleDataProvider:";
+        String errMsg2 = "org.ballerinalang.test.runtime.exceptions.BallerinaTestException: " +
+                "Error while invoking function 'testInvalidTupleDataProvider'";
+        String errMsg3 = "If you are using data providers please check if types return from data provider match" +
                 " test function parameter types";
         String[] args = mergeCoverageArgs(new String[]{"invalid-data-provider-test3.bal"});
         String output = balClient.runMainAndReadStdOut("test", args,
-                new HashMap<>(), projectPath, true);
-        if (!output.contains(errMsg) || !output.contains(errMsg2)) {
+                new HashMap<>(), projectPath, false);
+        if (!output.contains(errMsg1) || !output.contains(errMsg2) || !output.contains(errMsg3)) {
             AssertionUtils.assertForTestFailures(output, "error while handling invalid data providers");
         }
     }

